### DISCUSSION TITLE
Inquiry on Suggested Approaches for Centralized Management of Paths and Rights-of-Way in OpenTCS - Version 5.17.1

### DISCUSSION THREAD

**Discussion by BenYaho** [openTCS/opentcs#169] [Q&A]
Dear OpenTCS Team,

We are currently using OpenTCS version 5.17.1 for AGV system development. According to sections 4.2 and 4.3 of the User’s Guide, we have learned about functionalities related to path and resource management. Currently, the program first handles each vehicle’s path individually and then centrally manages the resources. Since we plan to implement centralized path and resource management for multiple vehicles, we would like to understand how to effectively extend path and resource management within the existing architecture.

Specifically, we have the following questions:

Path planning and resource management for centralized control: Are there recommended configurations or extension methods to improve collaboration between multiple vehicles, reduce path conflicts, and avoid deadlocks?

Considerations for extending or replacing core components: We are considering extending or replacing core Java classes like DefaultDispatcher, DefaultVehicleController, and AbstractReroutingStrategy. Do you have any suggestions for this? For example, how can we modify and extend these functions while ensuring system stability and safety?

We aim to implement these changes within the current framework to maintain system stability and safety during expansion. We look forward to your valuable advice and guidance.

**Comment by swltr** [comment]
@BenYaho I hope this helps:

* Centralized control: There is no explicit openTCS API for this so far. Routing is handled by the `Router` component, traffic management by the `Scheduler` component. It sounds like you would want to replace the default implementations for both and inject an additional shared custom component that manages a shared state (as a singleton, probably).
* Replacing core components: While the technical part of replacing these components via the Guice configuration is easy, I would recommend to do that only for components that are part of the openTCS API.
  * `AbstractReroutingStrategy` is not part of the API, and neither is the interface `ReroutingStrategy` which it implements. They both belong to the default dispatcher implementation, which may change with every openTCS release in ways that are not backward compatible.
  * `DefaultVehicleController` implements the interface `VehicleController`, which is part of the openTCS API. `DefaultVehicleController` itself is not, though, and may change with every openTCS release in ways that are not backward compatible.
  * `DefaultDispatcher`: as with `DefaultVehicleController`, the interface `Dispatcher` is part of the openTCS API, but `DefaultDispatcher` is not.

In general, the following rule applies: if your code depends on anything that is not part of the `opentcs-api-base` or `opentcs-api-injection` libraries, it is not safe with regards to compatibility between openTCS releases. It is recommended to code against these APIs only.

**Reply by cooldelay** [reply]
Dear openTCS team,

I am customizing DefaultVehicleControllerPool and VehicleControllerFactory to make the factory return an abstract VehicleController instead of the specific DefaultVehicleController. Our goal is to enable custom implementations of VehicleController based on varying requirements.

As shown in the attached image, we modified the factory to return a VehicleController. However, we encountered two errors:
No implementation for VehicleControllerFactory was bound.
VehicleController is an interface, not a concrete class. Unable to create AssistedInject factory.

Currently, the framework's factory only creates DefaultVehicleController instances. Could you provide guidance on how to modify or extend the factory to support creating different VehicleController implementations? Any best practices for achieving this within the existing framework would be appreciated.

![image](https://github.com/user-attachments/assets/23b658c9-672a-4e14-a759-9824f9916c5b)

**Reply by swltr** [reply]
@cooldelay Is this related to @BenYaho's original post? If not, please open a new discussion in the forum.

If by "customizing" you mean you're modifying `VehicleControllerFactory` and `DefaultVehicleControllerPool` themselves, you should be aware that this is not the intended way to integrate openTCS with your use case. You should program against the openTCS API, not modify the openTCS source code itself. Furthermore, `VehicleControllerFactory` and `DefaultVehicleControllerPool` are also not part of the openTCS API, so e.g. subclassing `DefaultVehicleControllerPool` is also not supported. What is part of the API is the interface `VehicleControllerPool`, so one possible way to achieve what you're trying to do is to bind your own implementation of `VehicleControllerPool` and work with that.
